src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/ProjectConfigurationStateManager.cs

```diff
@@ -0,0 +1,181 @@
+    private static readonly TimeSpan s_enqueueDelay = TimeSpan.FromMilliseconds(250);
+
+    public ProjectConfigurationStateManager(
+        IRazorProjectService projectService,
+        ILoggerFactory loggerFactory,
+        IProjectSnapshotManager projectManager)
+        : this(projectService,
+               loggerFactory,
+               projectManager,
+               s_enqueueDelay)
+    { }
+
+    // Provided for tests to specify enqueue delay
+    public ProjectConfigurationStateManager(
+        IRazorProjectService projectService,
+        ILoggerFactory loggerFactory,
+        IProjectSnapshotManager projectManager,
+        TimeSpan enqueueDelay)
+    {
+        _projectService = projectService;
+        _projectManager = projectManager;
+        _logger = loggerFactory.GetOrCreateLogger<ProjectConfigurationStateManager>();
+
+        _disposalTokenSource = new ();
+        _workQueue = new(
+            enqueueDelay,
+            ProcessBatchAsync,
+            _disposalTokenSource.Token);
+    }
+
+    public void Dispose()
+    {
+        _disposalTokenSource.Cancel();
+        _disposalTokenSource.Dispose();
+    }
+
+    private ValueTask ProcessBatchAsync(ImmutableArray<(ProjectKey ProjectKey, RazorProjectInfo? ProjectInfo)> workItems, CancellationToken cancellationToken)
+    {
+        foreach (var workItem in workItems.GetMostRecentUniqueItems(Comparer.Instance))
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return default;
+            }
+
+            UpdateProject(workItem.ProjectKey, workItem.ProjectInfo, cancellationToken);
+        }
+
+        return default;
+    }
+
+    public void ProjectInfoUpdated(ProjectKey projectKey, RazorProjectInfo? projectInfo, CancellationToken cancellationToken)
+    {
+        var knownProject = _projectManager.TryGetLoadedProject(projectKey, out _);
+
+        if (projectInfo is null)
+        {
+            if (knownProject)
+            {
+                // projectInfo null is "remove"
+                EnqueueUpdateProject(projectKey, projectInfo: null);
+            }
+            else
+            {
+                _logger.LogWarning($"Found no existing project key '{projectKey.Id}' but projectInfo is null. Assuming no-op deletion.");
+            }
+
+            return;
+        }
+
+        if (!knownProject)
+        {
+            // For the new code path using endpoint, SerializedFilePath is IntermediateOutputPath.
+            // We will rename it when we remove the old code path that actually uses serialized bin file.
+            var intermediateOutputPath = FilePathNormalizer.Normalize(projectInfo.SerializedFilePath);
+            _logger.LogInformation($"Found no existing project key for project key '{projectKey.Id}'. Assuming new project.");
+
+            AddProjectAsync(
+                intermediateOutputPath,
+                projectInfo,
+                cancellationToken).Forget();
+            return;
```

Comment:
Feels like the logic here with the early returns could be simplified. All code paths lead to `EnqueueUpdateProject`, except an unknown project with a null info, but one of them is hidden in this AddProject call, which is a fire-and-forget but seemingly only for code flow reasons.

Is it possible to make this method async, and have the flow something like:

```
if (!_projectManager.TryGetProject(...))
{
	if (projectInfo is not null)
        await AddProjectAsync
   else
        return; // no-op
}

EnqueueUpdate(projectKey, projectInfo)
```

Reply:
Agreed, refactored as suggested, thanks!